### ISSUE TITLE
Support bridged methods in ReflectionUtils.findAllMethodsInHierarchy()

### ISSUE THREAD

**Issue by sbrannen** [junit-team/junit5#333]
- [x] Support bridged methods in `ReflectionUtils.findAllMethodsInHierarchy(Class<?>, MethodSortOrder)`.
- [x] Ensure the Jupiter engine behaves correctly for bridge methods (see junit-team/junit4#1411).

**Comment by sormuras** [comment]
in progress

**Comment by sbrannen** [comment]
Reopening this issue.

Adding `@Test` to `org.junit.jupiter.engine.bridge.AbstractNumberTests.test(N)` causes the _overridden_ test method to be executed alongside the overriding test method in subclasses.

**Comment by sormuras** [comment]
This is due to a bug in `ReflectionUtils#isMethodShadowedBy(Method upper, Method lower)`. It should check for whether the signature of one method is a subsignature of another as specified here: https://docs.oracle.com/javase/specs/jls/se8/html/jls-8.html#jls-8.4.2

**Comment by sbrannen** [comment]
Reopening due to a regression: Code introduced in conjunction with this issue breaks standard support for test class hierarchies _without_ bridge or synthetic methods present.

Specifically, given the following JUnit Jupiter test cases, it is required that execution of `ByteTests` result in execution of `AbstractNumberTests.test(Number)` **and** `ByteTests.test(Byte)`. However, only the latter gets executed against the status quo on `master` since the current code in `ReflectionUtils.isMethodShadowedBy(Method, Method)` incorrectly thinks that `ByteTests.test(Byte)` _shadows_ `AbstractNumberTests.test(Number)`.

```java
class NumberResolver implements ParameterResolver {

	@Override
	public boolean supports(ParameterContext parameterContext, ExtensionContext extensionContext)
			throws ParameterResolutionException {

		return Number.class.isAssignableFrom(parameterContext.getParameter().getType());
	}

	@Override
	public Object resolve(ParameterContext parameterContext, ExtensionContext extensionContext)
			throws ParameterResolutionException {

		Class<?> type = parameterContext.getParameter().getType();
		try {
			if (type == Number.class) {
				return new Integer(42);
			}
			return type.getMethod("valueOf", String.class).invoke(null, "123");
		}
		catch (Exception e) {
			throw new AssertionError("Could not resolve number type: " + type, e);
		}
	}
}

@ExtendWith(NumberResolver.class)
public abstract class AbstractNumberTests {

	@Test
	public void test(Number value) {
		System.out.println("Number: " + value);
	}
}

public class ByteTests extends AbstractNumberTests {

	@Test
	public void test(Byte value) {
		System.out.println("Byte: " + value);
	}
}

````

**Comment by sbrannen** [comment]
Added _bug_ label until the current regression is fixed.

**Comment by sbrannen** [comment]
@sormuras, you fixed the regression -- right?

If so, please remove the _bug_ label and close this issue.

Cheers!

**Comment by sbrannen** [comment]
Fixed in: 0eccc25520e45310ef75295e4fcd0b221e4e556a -- right?

**Comment by sormuras** [comment]
Right. Removing the label(s). Slante!
